pkg/tekton/metapipeline/metaclient/metaclient.go

```diff
@@ -0,0 +1,324 @@
+package metaclient
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/jenkins-x/jx/pkg/apps"
+	jxclient "github.com/jenkins-x/jx/pkg/client/clientset/versioned"
+	"github.com/jenkins-x/jx/pkg/config"
+	"github.com/jenkins-x/jx/pkg/gits"
+	"github.com/jenkins-x/jx/pkg/jenkinsfile"
+	"github.com/jenkins-x/jx/pkg/jxfactory"
+	"github.com/jenkins-x/jx/pkg/kube"
+	"github.com/jenkins-x/jx/pkg/log"
+	"github.com/jenkins-x/jx/pkg/prow"
+	"github.com/jenkins-x/jx/pkg/tekton"
+	"github.com/jenkins-x/jx/pkg/tekton/metapipeline"
+	"github.com/jenkins-x/jx/pkg/tekton/syntax"
+	"github.com/jenkins-x/jx/pkg/util"
+	"github.com/pkg/errors"
+	tektonclient "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	corev1 "k8s.io/api/core/v1"
+	kubeclient "k8s.io/client-go/kubernetes"
+)
+
+const (
+	pullRefOptionName   = "pull-refs"
+	sourceURLOptionName = "source-url"
+
+	retryDuration      = time.Second * 30
+	defaultCheckoutDir = "source"
+)
+
+// MetaClient contains the arguments to create the meta pipeline
+type MetaClient struct {
+	SourceURL      string
+	PullRefs       string
+	Context        string
+	ServiceAccount string
+	CustomLabels   []string
+	CustomEnvs     []string
+	DefaultImage   string
+
+	Results     tekton.CRDWrapper
+	OutDir      string
+	NoApply     bool
+	versionsDir string
+	gitter      gits.Gitter
+	factory     jxfactory.Factory
+}
+
+// Create creates the meta pipeline
+func (c *MetaClient) Create() error {
```

Comment:
From the Create method, I would expect that it actually returns the created resources instead of applying them directly. This way a client can decide what to do with the resources. For example, apply them to the cluster or write them to disk or generate some shell script for the local execution of the commands in the pipeline (aka interpret mode).